File: funnyvote/app/src/main/java/com/heaton/funnyvote/ui/main/MainPageTabFragment.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.heaton.funnyvote.ui.main;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.widget.RelativeLayout;

import com.getbase.floatingactionbutton.FloatingActionButton;
import com.google.android.gms.analytics.Tracker;
import com.heaton.funnyvote.FunnyVoteApplication;
import com.heaton.funnyvote.R;
import com.heaton.funnyvote.data.VoteData.VoteDataRepository;
import com.heaton.funnyvote.database.User;
import com.heaton.funnyvote.database.VoteData;
import com.heaton.funnyvote.ui.HidingScrollListener;

import java.util.List;

import jp.wasabeef.recyclerview.adapters.ScaleInAnimationAdapter;

public class MainPageTabFragment extends Fragment implements MainPageContract.TabPageFragment {
    private static final int LIMIT = VoteDataRepository.PAGE_COUNT;
    public static String TAG = MainPageTabFragment.class.getSimpleName();

    public static final String KEY_TAB = "tab";
    public static final String KEY_LOGIN_USER = "key_login_user";
    public static final String KEY_TARGET_USER = "key_target_user";

    public static final String TAB_HOT = "HOT";
    public static final String TAB_NEW = "NEW";

    public static final String TAB_CREATE = "CREATE";
    public static final String TAB_PARTICIPATE = "PARTICIPATE";
    public static final String TAB_FAVORITE = "FAVORITE";

    private RecyclerView ryMain;
    private RelativeLayout RootView;
    private String tab = TAB_HOT;
    private List<VoteData> voteDataList;
    private VoteWallItemAdapter adapter;
    private SwipeRefreshLayout swipeRefreshLayout;
    private FloatingActionButton fabTop;
    private Tracker tracker;
    private MainPageContract.Presenter presenter;
    private VoteWallItemListener wallItemListener;

    public static MainPageTabFragment newInstance(String tab, User loginUser) {
        return newInstance(tab, loginUser, null);
    }

    public static MainPageTabFragment newInstance(String tab, User loginUser, User targetUser) {
        MainPageTabFragment fragment = new MainPageTabFragment();
        Bundle argument = new Bundle();
        argument.putString(MainPageTabFragment.KEY_TAB, tab);
        argument.putParcelable(MainPageTabFragment.KEY_LOGIN_USER, loginUser);
        argument.putParcelable(MainPageTabFragment.KEY_TARGET_USER, targetUser);
        fragment.setArguments(argument);
        fragment.setRetainInstance(false);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        Bundle argument = getArguments();
        this.tab = argument.getString(KEY_TAB);
        FunnyVoteApplication application = (FunnyVoteApplication) getActivity().getApplication();
        tracker = application.getDefaultTracker();
        RootView = (RelativeLayout) inflater.inflate(R.layout.fragment_main_page_tab, container, false);
        fabTop = (FloatingActionButton) RootView.findViewById(R.id.fabTop);
        fabTop.setVisibility(View.GONE);
        ryMain = (RecyclerView) RootView.findViewById(R.id.ryMainPage);
        wallItemListener = new VoteWallItemListener() {
            @Override
            public void onVoteFavoriteChange(VoteData voteData) {
                Log.d(TAG, "onVoteFavoriteChange");
                presenter.favoriteVote(voteData);
            }

            @Override
            public void onVoteItemClick(VoteData voteData) {
                presenter.IntentToVoteDetail(voteData);
            }

            @Override
            public void onVoteAuthorClick(VoteData voteData) {
                presenter.IntentToAuthorDetail(voteData);
            }

            @Override
            public void onVoteShare(VoteData voteData) {
                presenter.IntentToShareDialog(voteData);
            }

            @Override
            public void onVoteQuickPoll(VoteData voteData, String optionCode) {
                presenter.pollVote(voteData, optionCode, null);
            }

            @Override
            public void onNoVoteCreateNew() {
                presenter.IntentToCreateVote();
            }

            @Override
            public void onReloadVote() {
                if (tab.equals(TAB_HOT)) {
                    presenter.refreshHotList();
                } else if (tab.equals(TAB_NEW)) {
                    presenter.refreshNewList();
                } else if (tab.equals(TAB_CREATE)) {
                    presenter.refreshCreateList();
                } else if (tab.equals(TAB_PARTICIPATE)) {
                    presenter.refreshParticipateList();
                } else if (tab.equals(TAB_FAVORITE)) {
                    presenter.refreshFavoriteList();
                }
            }
        };
        swipeRefreshLayout = (SwipeRefreshLayout) RootView.findViewById(R.id.swipeLayout);
        swipeRefreshLayout.setOnRefreshListener(new WallItemOnRefreshListener());

        return RootView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        this.setPresenter(presenter);
        if (presenter == null) {
            return;
        }
        if (tab.equals(TAB_HOT)) {
            presenter.setHotsFragmentView(this);
        } else if (tab.equals(TAB_NEW)) {
            presenter.setNewsFragmentView(this);
        } else if (tab.equals(TAB_CREATE)) {
            presenter.setCreateFragmentView(this);
        } else if (tab.equals(TAB_PARTICIPATE)) {
            presenter.setParticipateFragmentView(this);
        } else if (tab.equals(TAB_FAVORITE)) {
            presenter.setFavoriteFragmentView(this);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        LinearLayoutManager manager = (LinearLayoutManager) ryMain.getLayoutManager();
        int position = manager.findFirstVisibleItemPosition();
        if (position == 0) {
            // TODO:AUTO UPDATE .
            //refreshData();
        }
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
    }

    @Override
    public void setUpRecycleView(List<VoteData> voteDataList) {
        //voteDataList = voteDataList;
        adapter = new VoteWallItemAdapter(getContext(), wallItemListener
                , voteDataList);
        // if max count is -1 , the list is init.
        adapter.setMaxCount(-1);
        if (tab.equals(TAB_HOT)) {
            adapter.setNoVoteTag(VoteWallItemAdapter.TAG_NO_VOTE_REFRESH);
        } else if (tab.equals(TAB_NEW)) {
            adapter.setNoVoteTag(VoteWallItemAdapter.TAG_NO_VOTE_REFRESH);
        } else if (tab.equals(TAB_CREATE)) {
            adapter.setNoVoteTag(VoteWallItemAdapter.TAG_NO_VOTE_CREATE_NEW);
        } else if (tab.equals(TAB_PARTICIPATE)) {
            adapter.setNoVoteTag(VoteWallItemAdapter.TAG_NO_VOTE_PARTICIPATE);
        } else if (tab.equals(TAB_FAVORITE)) {
            adapter.setNoVoteTag(VoteWallItemAdapter.TAG_NO_VOTE_FAVORITE);
        }
        adapter.resetItemTypeList();
        ScaleInAnimationAdapter scaleInAnimationAdapter = new ScaleInAnimationAdapter(adapter);
        scaleInAnimationAdapter.setDuration(1000);
        ryMain.setAdapter(adapter);
        ryMain.addOnScrollListener(new HidingScrollListener() {
            @Override
            public void onHide() {
                fabTop.animate().translationY(
                        fabTop.getHeight() + 50)
                        .setInterpolator(new AccelerateInterpolator(2));
            }

            @Override
            public void onShow() {
                this.resetScrollDistance();
                fabTop.animate().translationY(0).setInterpolator(new DecelerateInterpolator(2));
            }
        });
        fabTop.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                LinearLayoutManager manager = (LinearLayoutManager) ryMain.getLayoutManager();
                int position = manager.findFirstVisibleItemPosition();
                if (position > 5) {
                    ryMain.scrollToPosition(5);
                }
                ryMain.smoothScrollToPosition(0);
            }
        });
    }

    @Override
    public void refreshFragment(List<VoteData> voteDataList) {
        adapter.setVoteList(voteDataList);
        adapter.resetItemTypeList();
        adapter.notifyDataSetChanged();
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
    }

    @Override
    public void setPresenter(MainPageContract.Presenter presenter) {
        this.presenter = presenter;
    }

    private class WallItemOnRefreshListener implements SwipeRefreshLayout.OnRefreshListener {

        @Override
        public void onRefresh() {

            if (tab.equals(TAB_HOT)) {
                presenter.reloadHotList(0);
            } else if (tab.equals(TAB_NEW)) {
                presenter.reloadNewList(0);
            } else if (tab.equals(TAB_CREATE)) {
                presenter.reloadCreateList(0);
            } else if (tab.equals(TAB_PARTICIPATE)) {
                presenter.reloadParticipateList(0);
            } else if (tab.equals(TAB_FAVORITE)) {
                presenter.reloadFavoriteList(0);
            }
        }
    }


    @Override
    public void setMaxCount(int max) {
        if (adapter != null) {
            adapter.setMaxCount(max);
        }
    }

    @Override
    public void setTab(String tab) {
        this.tab = tab;
    }

    @Override
    public void hideSwipeLoadView() {
        if (swipeRefreshLayout.isRefreshing()) {
            swipeRefreshLayout.setRefreshing(false);
        }
    }

    public interface VoteWallItemListener {
        void onVoteFavoriteChange(VoteData voteData);

        void onVoteItemClick(VoteData voteData);

        void onVoteAuthorClick(VoteData voteData);

        void onVoteShare(VoteData voteData);

        void onVoteQuickPoll(VoteData voteData, String optionCode);

        void onNoVoteCreateNew();

        void onReloadVote();
    }
}

File: funnyvote/app/src/main/java/com/heaton/funnyvote/ui/main/MainPageTabFragment.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.heaton.funnyvote.ui.main

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateInterpolator
import android.view.animation.DecelerateInterpolator
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.gms.analytics.Tracker
import com.heaton.funnyvote.FunnyVoteApplication
import com.heaton.funnyvote.data.VoteData.VoteDataRepository
import com.heaton.funnyvote.database.User
import com.heaton.funnyvote.database.VoteData
import com.heaton.funnyvote.databinding.FragmentMainPageTabBinding
import com.heaton.funnyvote.ui.HidingScrollListener
import jp.wasabeef.recyclerview.adapters.ScaleInAnimationAdapter

class MainPageTabFragment : Fragment(), MainPageContract.TabPageFragment {

    private var _binding: FragmentMainPageTabBinding? = null
    private val binding get() = _binding!!

    private var tab: String = TAB_HOT
    private var adapter: VoteWallItemAdapter? = null
    private var tracker: Tracker? = null
    private lateinit var presenter: MainPageContract.Presenter
    private lateinit var wallItemListener: VoteWallItemListener

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentMainPageTabBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val argument = arguments
        this.tab = argument!!.getString(KEY_TAB) ?: TAB_HOT
        val application = requireActivity().application as FunnyVoteApplication
        tracker = application.defaultTracker
        binding.fabTop.visibility = View.GONE
        wallItemListener = object : VoteWallItemListener {
            override fun onVoteFavoriteChange(voteData: VoteData) = presenter.favoriteVote(voteData)

            override fun onVoteItemClick(voteData: VoteData) =
                    presenter.IntentToVoteDetail(voteData)

            override fun onVoteAuthorClick(voteData: VoteData) =
                    presenter.IntentToAuthorDetail(voteData)

            override fun onVoteShare(voteData: VoteData) = presenter.IntentToShareDialog(voteData)

            override fun onVoteQuickPoll(voteData: VoteData, optionCode: String) =
                    presenter.pollVote(voteData, optionCode, "")

            override fun onNoVoteCreateNew() = presenter.IntentToCreateVote()

            override fun onReloadVote() {
                when (tab) {
                    TAB_HOT -> presenter.refreshHotList()
                    TAB_NEW -> presenter.refreshNewList()
                    TAB_CREATE -> presenter.refreshCreateList()
                    TAB_PARTICIPATE -> presenter.refreshParticipateList()
                    TAB_FAVORITE -> presenter.refreshFavoriteList()
                }
            }
        }
        binding.swipeLayout.setOnRefreshListener(WallItemOnRefreshListener())
        this.setPresenter(presenter)
        when (tab) {
            TAB_HOT -> presenter.setHotsFragmentView(this)
            TAB_NEW -> presenter.setNewsFragmentView(this)
            TAB_CREATE -> presenter.setCreateFragmentView(this)
            TAB_PARTICIPATE -> presenter.setParticipateFragmentView(this)
            TAB_FAVORITE -> presenter.setFavoriteFragmentView(this)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onResume() {
        super.onResume()
        _binding?.let { b ->
            val manager = b.ryMainPage.layoutManager as? LinearLayoutManager
            val position = manager?.findFirstVisibleItemPosition() ?: 0
            if (position == 0) {
                // TODO:AUTO UPDATE
            }
        }
        adapter?.notifyDataSetChanged()
    }

    override fun setUpRecycleView(voteDataList: List<VoteData>) {
        val b = _binding ?: return
        adapter = VoteWallItemAdapter(requireContext(), wallItemListener, voteDataList)
        adapter!!.setMaxCount(-1)
        when (tab) {
            TAB_HOT -> adapter!!.setNoVoteTag(VoteWallItemAdapter.TAG_NO_VOTE_REFRESH)
            TAB_NEW -> adapter!!.setNoVoteTag(VoteWallItemAdapter.TAG_NO_VOTE_REFRESH)
            TAB_CREATE -> adapter!!.setNoVoteTag(VoteWallItemAdapter.TAG_NO_VOTE_CREATE_NEW)
            TAB_PARTICIPATE -> adapter!!.setNoVoteTag(VoteWallItemAdapter.TAG_NO_VOTE_PARTICIPATE)
            TAB_FAVORITE -> adapter!!.setNoVoteTag(VoteWallItemAdapter.TAG_NO_VOTE_FAVORITE)
        }
        adapter!!.resetItemTypeList()
        val itemAdapter = adapter ?: return
        val scaleInAnimationAdapter = ScaleInAnimationAdapter(itemAdapter)
        scaleInAnimationAdapter.setDuration(1000)
        b.ryMainPage.adapter = itemAdapter
        b.ryMainPage.addOnScrollListener(object : HidingScrollListener() {
            override fun onHide() {
                _binding?.let {
                    it.fabTop.animate().translationY(
                            (it.fabTop.height + 50).toFloat()).interpolator = AccelerateInterpolator(2f)
                }
            }

            override fun onShow() {
                this.resetScrollDistance()
                _binding?.let {
                    it.fabTop.animate().translationY(0f).interpolator = DecelerateInterpolator(2f)
                }
            }
        })
        b.fabTop.setOnClickListener {
            val manager = b.ryMainPage.layoutManager as? LinearLayoutManager
            val position = manager?.findFirstVisibleItemPosition() ?: 0
            if (position > 5) {
                b.ryMainPage.scrollToPosition(5)
            }
            b.ryMainPage.smoothScrollToPosition(0)
        }
    }

    override fun refreshFragment(voteDataList: List<VoteData>) {
        if (adapter != null) {
            adapter!!.setVoteList(voteDataList)
            adapter!!.resetItemTypeList()
            adapter!!.notifyDataSetChanged()
        }
    }

    override fun setPresenter(presenter: MainPageContract.Presenter) {
        this.presenter = presenter
    }

    private inner class WallItemOnRefreshListener : SwipeRefreshLayout.OnRefreshListener {
        override fun onRefresh() {
            when (tab) {
                TAB_HOT -> presenter.reloadHotList(0)
                TAB_NEW -> presenter.reloadNewList(0)
                TAB_CREATE -> presenter.reloadCreateList(0)
                TAB_PARTICIPATE -> presenter.reloadParticipateList(0)
                TAB_FAVORITE -> presenter.reloadFavoriteList(0)
            }
        }
    }

    override fun setMaxCount(max: Int) {
        adapter?.setMaxCount(max.toLong())
    }

    override fun setTab(tab: String) {
        this.tab = tab
    }

    override fun hideSwipeLoadView() {
        _binding?.let {
            if (it.swipeLayout.isRefreshing) {
                it.swipeLayout.isRefreshing = false
            }
        }
    }

    interface VoteWallItemListener {
        fun onVoteFavoriteChange(voteData: VoteData)
        fun onVoteItemClick(voteData: VoteData)
        fun onVoteAuthorClick(voteData: VoteData)
        fun onVoteShare(voteData: VoteData)
        fun onVoteQuickPoll(voteData: VoteData, optionCode: String)
        fun onNoVoteCreateNew()
        fun onReloadVote()
    }

    companion object {
        private val LIMIT = VoteDataRepository.PAGE_COUNT
        var TAG = MainPageTabFragment::class.java.simpleName

        const val KEY_TAB = "tab"
        const val KEY_LOGIN_USER = "key_login_user"
        const val KEY_TARGET_USER = "key_target_user"

        const val TAB_HOT = "HOT"
        const val TAB_NEW = "NEW"
        const val TAB_CREATE = "CREATE"
        const val TAB_PARTICIPATE = "PARTICIPATE"
        const val TAB_FAVORITE = "FAVORITE"

        @JvmOverloads
        fun newInstance(tab: String, loginUser: User, targetUser: User? = null): MainPageTabFragment {
            val fragment = MainPageTabFragment()
            val argument = Bundle()
            argument.putString(MainPageTabFragment.KEY_TAB, tab)
            argument.putParcelable(MainPageTabFragment.KEY_LOGIN_USER, loginUser)
            argument.putParcelable(MainPageTabFragment.KEY_TARGET_USER, targetUser)
            fragment.arguments = argument
            fragment.retainInstance = false
            return fragment
        }
    }
}
